system/lib/wasmfs/wasmfs.cpp

```diff
@@ -113,10 +116,71 @@ __wasi_errno_t __wasi_fd_read(
     const uint8_t* buf = iovs[i].buf;
     __wasi_size_t len = iovs[i].buf_len;
 
-    file.read(buf, len);
+    static_cast<DataFile::Handle&>(file).read(buf, len);
     num += len;
   }
   *nread = num;
   return __WASI_ERRNO_INVAL;
 }
+
+__wasi_fd_t __syscall_open(long pathname, long flags, long mode) {
+  // Obtain OpenFile from the open file table. Release fileTable lock using lambda function.
+  auto fileTable = [] { return FileTable::get(); }();
+
+  int accessMode = (flags & O_ACCMODE);
+  bool canWrite = false;
+
+  switch (accessMode) {
+    case O_RDONLY:
+      canWrite = false;
+    case O_WRONLY:
+      canWrite = true;
+    case O_RDWR:
+      canWrite = true;
+    default:
+      canWrite = false;
+  }
+
+  std::string interim = "";
+  std::vector<std::string> pathParts;
+
+  auto newPathName = (const char*)pathname;
+  for (int i = 0; i < strlen(newPathName); i++) {
+    if (newPathName[i] != '/') {
+      interim += newPathName[i];
+    } else {
+      pathParts.push_back(interim);
+      interim = "";
+    }
+  }
+
+  auto fileFromPath = RootDirectory::getSharedPtr();
+  for (int i = 0; i < pathParts.size(); i++) {
+
+    // If it is a file and we have not reached the end of the path, exit
+    if (fileFromPath->is<DataFile>() && i != pathParts.size() - 1) {
+      return -(EBADF);
+    }
+
+    // Find the next entry in the current directory entry
+    fileFromPath->get().printKeys();
+    fileFromPath = std::static_pointer_cast<Directory>(fileFromPath->get().getEntry(pathParts[i]));
```

Comment:
Definitely don't want to use `std::static_pointer_cast` here because it doesn't know about our casting scheme. You probably want `fileFromPath` to be a `std::shared_ptr<File>` instead to make the types work out.